File: app/src/main/java/com/example/daisyling/common/util/download/InputParameter.java
package com.example.daisyling.common.util.download;

/**
 * Created by Emily on 10/11/21
 */
public class InputParameter {
    private final String baseUrl;
    private final String relativeUrl;
    private final String loadedFilePath;
    private final boolean isCallbackOnUiThread;

    private InputParameter(Builder builder) {
        this.baseUrl = builder.baseUrl;
        this.relativeUrl = builder.relativeUrl;
        this.loadedFilePath = builder.loadedFilePath;
        this.isCallbackOnUiThread = builder.isCallbackOnUiThread;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public String getRelativeUrl() {
        return relativeUrl;
    }

    public String getLoadedFilePath() {
        return loadedFilePath;
    }

    public boolean isCallbackOnUiThread() {
        return isCallbackOnUiThread;
    }

    public static class Builder {
        String baseUrl;
        String relativeUrl;
        String loadedFilePath;
        boolean isCallbackOnUiThread;

        public Builder(String baseUrl, String relativeUrl, String loadedFilePath) {
            this.baseUrl = baseUrl;
            this.relativeUrl = relativeUrl;
            this.loadedFilePath = loadedFilePath;
        }

        /**
         * Callback on UI thread
         */
        public Builder setCallbackOnUiThread(boolean callbackOnUiThread) {
            isCallbackOnUiThread = callbackOnUiThread;
            return this;
        }

        public InputParameter build() {
            return new InputParameter(this);
        }
    }
}

File: app/src/main/java/com/example/daisyling/common/util/download/InputParameter.kt
package com.example.daisyling.common.util.download

/**
 * Created by Emily on 10/11/21
 */
class InputParameter private constructor(builder: Builder) {
    val baseUrl: String = builder.baseUrl
    val relativeUrl: String = builder.relativeUrl
    val loadedFilePath: String = builder.loadedFilePath
    val isCallbackOnUiThread: Boolean

    class Builder(var baseUrl: String, var relativeUrl: String, var loadedFilePath: String) {
        var isCallbackOnUiThread = false

        /**
         * Callback on UI thread
         */
        fun setCallbackOnUiThread(callbackOnUiThread: Boolean): Builder {
            isCallbackOnUiThread = callbackOnUiThread
            return this
        }

        fun build(): InputParameter {
            return InputParameter(this)
        }
    }

    init {
        isCallbackOnUiThread = builder.isCallbackOnUiThread
    }
}